oracle/stork_fetcher.go

```diff
@@ -0,0 +1,216 @@
+package oracle
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"cosmossdk.io/math"
+	"github.com/InjectiveLabs/metrics"
+	oracletypes "github.com/InjectiveLabs/sdk-go/chain/oracle/types"
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+	log "github.com/xlab/suplog"
+)
+
+const (
+	messageTypeInvalid      messageType = "invalid_message"
+	messageTypeOraclePrices messageType = "oracle_prices"
+	messageTypeSubscribe    messageType = "subscribe"
+)
+
+var ErrInvalidMessage = errors.New("received invalid message")
+
+type StorkFetcher interface {
+	Start(ctx context.Context, conn *websocket.Conn) error
+	AssetPair(ticker string) *oracletypes.AssetPair
+}
+
+type messageType string
+
+func (m messageType) String() string {
+	return string(m)
+}
+
+type StorkConfig struct {
+	WebsocketUrl    string
+	WebsocketHeader string
+	Message         string
+}
+
+type storkFetcher struct {
+	conn        *websocket.Conn
+	latestPairs map[string]*oracletypes.AssetPair
+	tickers     []string
+	message     string
+	mu          sync.RWMutex
+
+	logger  log.Logger
+	svcTags metrics.Tags
+}
+
+// NewStorkFetcher returns a new StorkFetcher instance.
+func NewStorkFetcher(storkMessage string, storkTickers []string) *storkFetcher {
+	feed := &storkFetcher{
+		message:     storkMessage,
+		tickers:     storkTickers,
+		latestPairs: make(map[string]*oracletypes.AssetPair),
+		logger: log.WithFields(log.Fields{
+			"svc":      "oracle",
+			"dynamic":  true,
+			"provider": "storkFetcher",
+		}),
+
+		svcTags: metrics.Tags{
+			"provider": "storkFetcher",
+		},
+	}
+
+	return feed
+}
+
+func (f *storkFetcher) AssetPair(ticker string) *oracletypes.AssetPair {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	return f.latestPairs[ticker]
+}
+
+func (f *storkFetcher) Start(_ context.Context, conn *websocket.Conn) error {
+	f.conn = conn
+
+	err := f.subscribe()
+	if err != nil {
+		return err
+	}
+
```

Comment:
maybe defering here the reset would make the code in startReadingMessages clearer

Reply:
totally right!